src/nbl/asset/utils/ISPIRVDebloater.cpp

```diff
@@ -0,0 +1,115 @@
+            case  spv::ExecutionModelGLCompute : return hlsl::ESS_COMPUTE;
+            case  spv::ExecutionModelTaskEXT : return hlsl::ESS_TASK;
+            case  spv::ExecutionModelMeshEXT: return hlsl::ESS_MESH;
+            case  spv::ExecutionModelRayGenerationKHR : return hlsl::ESS_RAYGEN;
+            case  spv::ExecutionModelAnyHitKHR : return hlsl::ESS_ANY_HIT;
+            case  spv::ExecutionModelClosestHitKHR : return hlsl::ESS_CLOSEST_HIT;
+            case  spv::ExecutionModelMissKHR : return hlsl::ESS_MISS;
+            case  spv::ExecutionModelIntersectionKHR : return hlsl::ESS_INTERSECTION;
+            case  spv::ExecutionModelCallableKHR : return hlsl::ESS_CALLABLE;
+            default:
+            {
+                assert(false);
+                return hlsl::ESS_UNKNOWN;
+            }
+        }
+      };
+    static constexpr auto HEADER_SIZE = 5;
+
+    core::vector<bool> foundEntryPoints(entryPoints.size(), false);
+    std::vector<uint32_t> minimizedSpirv;
+    bool needDebloat = false;
+
+    auto offset = HEADER_SIZE;
+
+    while (offset < spirvDwordCount) {
+        const auto curOffset = offset;
+        const auto instruction = spirv[curOffset];
+
+        const auto length = instruction >> 16;
+        const auto opcode = instruction & 0x0ffffu;
+
+        offset += length;
+
+        if (opcode == spv::OpEntryPoint)
+        {
+            const auto curExecutionModel = static_cast<spv::ExecutionModel>(spirv[curOffset + 1]);
+
+            // TODO: check whether this reinterpret_cast is UB
+            const auto curEntryPointName = std::string_view(reinterpret_cast<const char*>(spirv + curOffset + 3)); // entryPoint name is the third parameter
+
+            auto findEntryPointIt = std::find(entryPoints.begin(), entryPoints.end(), 
+              EntryPoint{
+                .name = curEntryPointName,
+                .shaderStage = getHlslShaderStage(curExecutionModel),
+              });
+            if (findEntryPointIt != entryPoints.end())
+            {
+                foundEntryPoints[std::distance(findEntryPointIt, entryPoints.begin())] = true;
+            } else
+            {
+                if (needDebloat == false)
+                {
+                    minimizedSpirv.reserve(spirvDwordCount);
+                    minimizedSpirv.insert(minimizedSpirv.end(), spirv, spirv + curOffset);
+                    needDebloat = true;
+                }
+                continue;
+            }
+        }
+        
+        if (!needDebloat) continue;
+        minimizedSpirv.insert(minimizedSpirv.end(), spirv + curOffset, spirv + offset);
+    }
+
+    const auto isAllEntryPointsFound = std::all_of(foundEntryPoints.begin(), foundEntryPoints.end(), std::identity{});
+
+    if (!isAllEntryPointsFound || !needDebloat) {
+        return Result {
+            .isAllEntryPointsFound = isAllEntryPointsFound,
+            .spirv = nullptr,
+        };
+    }
```

Comment:
handle the no-debloat-needed success case separately from failure due to missing entry points